File: DSA-DAILY/28-07-2025/2044. Count Number of Maximum Bitwise-OR Subsets@leetcode-med.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    //this is a dynamic programming solution
    //we use a dp array where dp[i] represents the number of subsets that can achieve 
    int countMaxOrSubsets(vector<int>& nums) {
        int max=0;
        vector<int>dp(1<<17,0);
        dp[0]=1;
        for(int num:nums){
            for(int i=max;i>=0;i--){
                dp[i | num]+=dp[i];
            }
            max |=num;
        }
        return dp[max];
    }
};
//another approach is to use backtracking which i will implemented in other language
int main(){
    Solution sol;
    int n;
    cin >> n;
    vector<int> nums;
    for(int i = 0; i < n; i++){
        int x;
        cin >> x;
        nums.push_back(x);
    }
    cout << sol.countMaxOrSubsets(nums) << endl;
    return 0;
}

File: DSA-DAILY/28-07-2025/2044. Count Number of Maximum Bitwise-OR Subsets@leetcode-med.py
import random
from typing import List
class Solution:
    def countMaxOrSubsets(self, nums: List[int]) -> int:
        maxi=0
        for i in nums:
            maxi |=i
        return self.backtrack(nums,maxi,0,0)
    
    def backtrack(self,nums:List[int],maxi:int,ind:int,curr:int):
        if ind == len(nums):
            return 1 if curr == maxi else 0 
               
               
        if curr is maxi:
            return 1<<(len(nums)-ind)
        consider=self.backtrack(nums,maxi,ind+1,curr|nums[ind])
        notconsider=self.backtrack(nums,maxi,ind+1,curr)
        return consider+notconsider 
    
if __name__== "__main__":
    sol=Solution()
    nums=[3,1,5,2]
    print(sol.countMaxOrSubsets(nums))  